// @covalent/dynamic-forms/bundles/covalent-dynamic-forms.umd.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/forms'), require('@angular/cdk/portal'), require('rxjs/observable/timer'), require('rxjs/operator/toPromise'), require('@angular/common'), require('@angular/material/input'), require('@angular/material/select'), require('@angular/material/checkbox'), require('@angular/material/slider'), require('@angular/material/slide-toggle'), require('@angular/material/icon'), require('@angular/material/button'), require('@angular/material/datepicker'), require('@covalent/core/common'), require('@covalent/core/file')) :
	typeof define === 'function' && define.amd ? define(['exports', '@angular/core', '@angular/forms', '@angular/cdk/portal', 'rxjs/observable/timer', 'rxjs/operator/toPromise', '@angular/common', '@angular/material/input', '@angular/material/select', '@angular/material/checkbox', '@angular/material/slider', '@angular/material/slide-toggle', '@angular/material/icon', '@angular/material/button', '@angular/material/datepicker', '@covalent/core/common', '@covalent/core/file'], factory) :
	(factory((global.covalent = global.covalent || {}, global.covalent['dynamic-forms'] = {}),global.ng.core,global.ng.forms,global.ng.cdk.portal,global.Rx.Observable,global.Rx.Observable.prototype,global.ng.common,global.ng.material.input,global.ng.material.select,global.ng.material.checkbox,global.ng.material.slider,global.ng.material['slide-toggle'],global.ng.material.icon,global.ng.material.button,global.ng.material.datepicker,global.covalent.core.common,global.covalent.core.file));
}(this, (function (exports,core,forms,portal,timer,toPromise,common,input,select,checkbox,slider,slideToggle,icon,button,datepicker,common$1,file) { 'use strict';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the
License at http://www.apache.org/licenses/LICENSE-2.0
THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
MERCHANTABLITY OR NON-INFRINGEMENT.
See the Apache Version 2.0 License for specific language governing permissions
and limitations under the License.
***************************************************************************** */
/* global Reflect, Promise */
var extendStatics = Object.setPrototypeOf ||
    ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
    function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
function __extends(d, b) {
    extendStatics(d, b);
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
}









function __read(o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
}
function __spread() {
    for (var ar = [], i = 0; i < arguments.length; i++)
        ar = ar.concat(__read(arguments[i]));
    return ar;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var noop = function () {
    // empty method
};
/**
 * @abstract
 */
var AbstractControlValueAccessor = /** @class */ (function () {
    function AbstractControlValueAccessor() {
        /**
         * Implemented as part of ControlValueAccessor.
         */
        this._value = undefined;
        this.onChange = function (_) { return noop; };
        this.onTouched = function () { return noop; };
    }
    Object.defineProperty(AbstractControlValueAccessor.prototype, "value", {
        /**
         * @return {?}
         */
        get: function () { return this._value; },
        /**
         * @param {?} v
         * @return {?}
         */
        set: function (v) {
            if (v !== this._value) {
                this._value = v;
                this.onChange(v);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Implemented as part of ControlValueAccessor.
     * @param {?} value
     * @return {?}
     */
    AbstractControlValueAccessor.prototype.writeValue = function (value) {
        this.value = value;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    AbstractControlValueAccessor.prototype.registerOnChange = function (fn) {
        this.onChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    AbstractControlValueAccessor.prototype.registerOnTouched = function (fn) {
        this.onTouched = fn;
    };
    return AbstractControlValueAccessor;
}());
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var INPUT_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicInputComponent; }),
    multi: true,
};
var TdDynamicInputComponent = /** @class */ (function (_super) {
    __extends(TdDynamicInputComponent, _super);
    function TdDynamicInputComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.label = '';
        _this.type = undefined;
        _this.required = undefined;
        _this.min = undefined;
        _this.max = undefined;
        _this.minLength = undefined;
        _this.maxLength = undefined;
        return _this;
    }
    return TdDynamicInputComponent;
}(AbstractControlValueAccessor));
TdDynamicInputComponent.decorators = [
    { type: core.Component, args: [{
                providers: [INPUT_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-input',
                styles: [".td-dynamic-input-wrapper{\n  -webkit-box-orient:horizontal;\n  -webkit-box-direction:normal;\n      -ms-flex-direction:row;\n          flex-direction:row;\n  display:-webkit-box;\n  display:-ms-flexbox;\n  display:flex;\n  -webkit-box-sizing:border-box;\n          box-sizing:border-box; }\n  .td-dynamic-input-wrapper .td-dynamic-input-field{\n    -webkit-box-flex:1;\n        -ms-flex:1;\n            flex:1;\n    -webkit-box-sizing:border-box;\n            box-sizing:border-box; }\n"],
                template: "<div class=\"td-dynamic-input-wrapper\">\n  <mat-form-field class=\"td-dynamic-input-field\">\n    <input #elementInput\n            matInput\n            [(ngModel)]=\"value\"\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [type]=\"type\"\n            [required]=\"required\"\n            [attr.min]=\"min\"\n            [attr.max]=\"max\"\n            [attr.minLength]=\"minLength\"\n            [attr.maxLength]=\"maxLength\"/>\n  </mat-form-field>\n</div>\n",
            },] },
];
/** @nocollapse */
TdDynamicInputComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var UPLOAD_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicFileInputComponent; }),
    multi: true,
};
var TdDynamicFileInputComponent = /** @class */ (function (_super) {
    __extends(TdDynamicFileInputComponent, _super);
    function TdDynamicFileInputComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.required = undefined;
        _this.label = '';
        return _this;
    }
    return TdDynamicFileInputComponent;
}(AbstractControlValueAccessor));
TdDynamicFileInputComponent.decorators = [
    { type: core.Component, args: [{
                providers: [UPLOAD_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-file-input',
                styles: [".td-dynamic-file-input-wrapper{\n  -webkit-box-orient:horizontal;\n  -webkit-box-direction:normal;\n      -ms-flex-direction:row;\n          flex-direction:row;\n  display:-webkit-box;\n  display:-ms-flexbox;\n  display:flex;\n  -webkit-box-sizing:border-box;\n          box-sizing:border-box; }\n  .td-dynamic-file-input-wrapper .td-dynamic-file-input-field{\n    -webkit-box-flex:1;\n        -ms-flex:1;\n            flex:1;\n    -webkit-box-sizing:border-box;\n            box-sizing:border-box; }\n.td-file-input{\n  margin-left:10px; }\n"],
                template: "<div class=\"td-dynamic-file-input-wrapper\">\n  <mat-form-field tdFileDrop\n                  class=\"td-dynamic-file-input-field\"\n                  floatPlaceholder=\"never\"\n                  (fileDrop)=\"value = $event\"\n                  (click)=\"fileInput.inputElement.click()\"\n                  (keyup.enter)=\"fileInput.inputElement.click()\"\n                  (keyup.delete)=\"fileInput.clear()\"\n                  (keyup.backspace)=\"fileInput.clear()\">\n    <input matInput\n          [value]=\"value?.name\"\n          [placeholder]=\"label\"\n          readonly/>\n  </mat-form-field>\n  <button mat-icon-button *ngIf=\"value\" (click)=\"fileInput.clear()\" (keyup.enter)=\"fileInput.clear()\">\n    <mat-icon>cancel</mat-icon>\n  </button>\n  <td-file-input class=\"td-file-input\" #fileInput [(ngModel)]=\"value\">\n    <mat-icon>folder</mat-icon>\n    <span>{{ label }}</span>\n  </td-file-input>\n</div>",
            },] },
];
/** @nocollapse */
TdDynamicFileInputComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var TEXTAREA_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicTextareaComponent; }),
    multi: true,
};
var TdDynamicTextareaComponent = /** @class */ (function (_super) {
    __extends(TdDynamicTextareaComponent, _super);
    function TdDynamicTextareaComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.label = '';
        _this.required = undefined;
        return _this;
    }
    return TdDynamicTextareaComponent;
}(AbstractControlValueAccessor));
TdDynamicTextareaComponent.decorators = [
    { type: core.Component, args: [{
                providers: [TEXTAREA_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-textarea',
                styles: [".td-dynamic-textarea-wrapper{\n  -webkit-box-orient:horizontal;\n  -webkit-box-direction:normal;\n      -ms-flex-direction:row;\n          flex-direction:row;\n  display:-webkit-box;\n  display:-ms-flexbox;\n  display:flex;\n  -webkit-box-sizing:border-box;\n          box-sizing:border-box; }\n  .td-dynamic-textarea-wrapper .td-dynamic-textarea-field{\n    -webkit-box-flex:1;\n        -ms-flex:1;\n            flex:1;\n    -webkit-box-sizing:border-box;\n            box-sizing:border-box; }\n"],
                template: "<div class=\"td-dynamic-textarea-wrapper\">\n  <mat-form-field class=\"td-dynamic-textarea-field\">\n    <textarea #elementInput\n            matInput\n            [(ngModel)]=\"value\"\n            [placeholder]=\"label\"\n            [required]=\"required\"\n            rows=\"4\">\n    </textarea>\n  </mat-form-field>\n</div>",
            },] },
];
/** @nocollapse */
TdDynamicTextareaComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SLIDE_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicSlideToggleComponent; }),
    multi: true,
};
var TdDynamicSlideToggleComponent = /** @class */ (function (_super) {
    __extends(TdDynamicSlideToggleComponent, _super);
    function TdDynamicSlideToggleComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.label = '';
        _this.required = false;
        return _this;
    }
    return TdDynamicSlideToggleComponent;
}(AbstractControlValueAccessor));
TdDynamicSlideToggleComponent.decorators = [
    { type: core.Component, args: [{
                providers: [SLIDE_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-slide-toggle',
                styles: [""],
                template: "<div class=\"td-dynamic-slide-toggle-wrapper\">\n  <mat-slide-toggle [(ngModel)]=\"value\"\n                   [required]=\"required\">\n    {{label}}\n  </mat-slide-toggle>\n</div>",
            },] },
];
/** @nocollapse */
TdDynamicSlideToggleComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var CHECKBOX_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicCheckboxComponent; }),
    multi: true,
};
var TdDynamicCheckboxComponent = /** @class */ (function (_super) {
    __extends(TdDynamicCheckboxComponent, _super);
    function TdDynamicCheckboxComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.label = '';
        _this.required = false;
        return _this;
    }
    return TdDynamicCheckboxComponent;
}(AbstractControlValueAccessor));
TdDynamicCheckboxComponent.decorators = [
    { type: core.Component, args: [{
                providers: [CHECKBOX_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-checkbox',
                styles: [""],
                template: "<div class=\"td-dynamic-checkbox-wrapper\">\n  <mat-checkbox [(ngModel)]=\"value\"\n                [required]=\"required\">\n    {{label}}\n  </mat-checkbox>\n</div>",
            },] },
];
/** @nocollapse */
TdDynamicCheckboxComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SLIDER_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicSliderComponent; }),
    multi: true,
};
var TdDynamicSliderComponent = /** @class */ (function (_super) {
    __extends(TdDynamicSliderComponent, _super);
    function TdDynamicSliderComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.label = '';
        _this.required = undefined;
        _this.min = undefined;
        _this.max = undefined;
        return _this;
    }
    return TdDynamicSliderComponent;
}(AbstractControlValueAccessor));
TdDynamicSliderComponent.decorators = [
    { type: core.Component, args: [{
                providers: [SLIDER_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-slider',
                styles: [".td-dynamic-slider-field{\n  position:relative;\n  margin-top:8px;\n  -webkit-box-orient:horizontal;\n  -webkit-box-direction:normal;\n      -ms-flex-direction:row;\n          flex-direction:row;\n  display:-webkit-box;\n  display:-ms-flexbox;\n  display:flex;\n  -webkit-box-sizing:border-box;\n          box-sizing:border-box; }\n  .td-dynamic-slider-field .td-dynamic-slider{\n    -webkit-box-flex:1;\n        -ms-flex:1;\n            flex:1; }\n"],
                template: "<div class=\"td-dynamic-slider-wrapper\">\n  <div class=\"mat-form-field-placeholder-wrapper mat-form-field-can-float mat-form-field-should-float\"\n       [class.mat-focused]=\"slider._isActive\">\n    <label class=\"mat-form-field-placeholder mat-float mat-form-field-float td-slider-label\"> {{label}} <span *ngIf=\"required\" class=\"mat-placeholder-required\">*</span></label>\n  </div>\n  <div class=\"td-dynamic-slider-field\">\n    <mat-slider #slider\n               class=\"td-dynamic-slider\"\n               [(ngModel)]=\"value\"\n                [min]=\"min\"\n                [max]=\"max\"\n                thumbLabel\n                tickInterval=\"auto\"\n                [required]=\"required\">\n    </mat-slider>\n  </div>\n</div>",
            },] },
];
/** @nocollapse */
TdDynamicSliderComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var SELECT_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicSelectComponent; }),
    multi: true,
};
var TdDynamicSelectComponent = /** @class */ (function (_super) {
    __extends(TdDynamicSelectComponent, _super);
    function TdDynamicSelectComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.label = '';
        _this.required = undefined;
        _this.selections = undefined;
        return _this;
    }
    return TdDynamicSelectComponent;
}(AbstractControlValueAccessor));
TdDynamicSelectComponent.decorators = [
    { type: core.Component, args: [{
                providers: [SELECT_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-select',
                styles: [".td-dynamic-select-wrapper{\n  -webkit-box-orient:horizontal;\n  -webkit-box-direction:normal;\n      -ms-flex-direction:row;\n          flex-direction:row;\n  display:-webkit-box;\n  display:-ms-flexbox;\n  display:flex;\n  -webkit-box-sizing:border-box;\n          box-sizing:border-box; }\n  .td-dynamic-select-wrapper .td-dynamic-select-field{\n    -webkit-box-flex:1;\n        -ms-flex:1;\n            flex:1;\n    -webkit-box-sizing:border-box;\n            box-sizing:border-box; }\n"],
                template: "<div class=\"td-dynamic-select-wrapper\">\n  <mat-form-field class=\"td-dynamic-select-field\">\n    <mat-select [(ngModel)]=\"value\"\n              [placeholder]=\"label\"\n              [required]=\"required\">\n      <mat-option *ngFor=\"let selection of selections\" [value]=\"selection.value || selection\">{{selection.label || selection}}</mat-option>\n    </mat-select>\n  </mat-form-field>\n</div>\n",
            },] },
];
/** @nocollapse */
TdDynamicSelectComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var DATEPICKER_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicDatepickerComponent; }),
    multi: true,
};
var TdDynamicDatepickerComponent = /** @class */ (function (_super) {
    __extends(TdDynamicDatepickerComponent, _super);
    function TdDynamicDatepickerComponent() {
        var _this = _super.apply(this, __spread(arguments)) || this;
        _this.label = '';
        _this.type = undefined;
        _this.required = undefined;
        _this.min = undefined;
        _this.max = undefined;
        return _this;
    }
    return TdDynamicDatepickerComponent;
}(AbstractControlValueAccessor));
TdDynamicDatepickerComponent.decorators = [
    { type: core.Component, args: [{
                providers: [DATEPICKER_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-datepicker',
                styles: [".td-dynamic-datepicker-wrapper{\n  -webkit-box-orient:horizontal;\n  -webkit-box-direction:normal;\n      -ms-flex-direction:row;\n          flex-direction:row;\n  display:-webkit-box;\n  display:-ms-flexbox;\n  display:flex;\n  -webkit-box-sizing:border-box;\n          box-sizing:border-box; }\n  .td-dynamic-datepicker-wrapper .td-dynamic-datepicker-field{\n    -webkit-box-flex:1;\n        -ms-flex:1;\n            flex:1;\n    -webkit-box-sizing:border-box;\n            box-sizing:border-box; }\n"],
                template: "<div class=\"td-dynamic-datepicker-wrapper\">\n  <mat-form-field class=\"td-dynamic-datepicker-field\">\n    <input #elementInput\n            matInput\n            [matDatepicker]=\"dynamicDatePicker\"\n            [(ngModel)]=\"value\"\n            [formControl]=\"control\"\n            [placeholder]=\"label\"\n            [required]=\"required\"\n            [min]=\"min\"\n            [max]=\"max\"/>\n    <mat-datepicker-toggle matSuffix [for]=\"dynamicDatePicker\"></mat-datepicker-toggle>\n    <mat-datepicker #dynamicDatePicker></mat-datepicker>\n  </mat-form-field>\n</div>\n",
            },] },
];
/** @nocollapse */
TdDynamicDatepickerComponent.ctorParameters = function () { return []; };
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/** @enum {string} */
var TdDynamicType = {
    Text: 'text',
    Boolean: 'boolean',
    Number: 'number',
    Array: 'array',
    Date: 'date',
};
/** @enum {string} */
var TdDynamicElement = {
    Input: 'input',
    Datepicker: 'datepicker',
    Password: 'password',
    Textarea: 'textarea',
    Slider: 'slider',
    SlideToggle: 'slide-toggle',
    Checkbox: 'checkbox',
    Select: 'select',
    FileInput: 'file-input',
};
/**
 * @record
 */
/**
 * @record
 */
var DYNAMIC_ELEMENT_NAME_REGEX = /^[a-zA-Z]+[a-zA-Z0-9-_]*$/;
var TdDynamicFormsService = /** @class */ (function () {
    function TdDynamicFormsService() {
    }
    /**
     * Method to validate if the [name] is a proper element name.
     * Throws error if name is not valid.
     * @param {?} name
     * @return {?}
     */
    TdDynamicFormsService.prototype.validateDynamicElementName = function (name) {
        if (!DYNAMIC_ELEMENT_NAME_REGEX.test(name)) {
            throw new Error("Dynamic element name: \"" + name + "\" is not valid.");
        }
    };
    /**
     * Gets component to be rendered depending on [TdDynamicElement | TdDynamicType]
     * Throws error if it does not exists or not supported.
     * @param {?} element
     * @return {?}
     */
    TdDynamicFormsService.prototype.getDynamicElement = function (element) {
        switch (element) {
            case TdDynamicType.Text:
            case TdDynamicType.Number:
            case TdDynamicElement.Input:
            case TdDynamicElement.Password:
                return TdDynamicInputComponent;
            case TdDynamicElement.Textarea:
                return TdDynamicTextareaComponent;
            case TdDynamicType.Boolean:
            case TdDynamicElement.SlideToggle:
                return TdDynamicSlideToggleComponent;
            case TdDynamicElement.Checkbox:
                return TdDynamicCheckboxComponent;
            case TdDynamicElement.Slider:
                return TdDynamicSliderComponent;
            case TdDynamicType.Array:
            case TdDynamicElement.Select:
                return TdDynamicSelectComponent;
            case TdDynamicElement.FileInput:
                return TdDynamicFileInputComponent;
            case TdDynamicElement.Datepicker:
            case TdDynamicType.Date:
                return TdDynamicDatepickerComponent;
            default:
                throw new Error("Error: type " + element + " does not exist or not supported.");
        }
    };
    /**
     * Creates form control for element depending [ITdDynamicElementConfig] properties.
     * @param {?} config
     * @return {?}
     */
    TdDynamicFormsService.prototype.createFormControl = function (config) {
        var /** @type {?} */ validator = this.createValidators(config);
        return new forms.FormControl(config.default, validator);
    };
    /**
     * Creates form validationdepending [ITdDynamicElementConfig] properties.
     * @param {?} config
     * @return {?}
     */
    TdDynamicFormsService.prototype.createValidators = function (config) {
        var /** @type {?} */ validator;
        if (config.required) {
            validator = forms.Validators.required;
        }
        if (config.max || config.max === 0) {
            validator = forms.Validators.compose([validator, forms.Validators.max(parseFloat(config.max))]);
        }
        if (config.min || config.min === 0) {
            validator = forms.Validators.compose([validator, forms.Validators.min(parseFloat(config.min))]);
        }
        if (config.maxLength || config.maxLength === 0) {
            validator = forms.Validators.compose([validator, forms.Validators.maxLength(parseFloat(config.maxLength))]);
        }
        if (config.minLength || config.minLength === 0) {
            validator = forms.Validators.compose([validator, forms.Validators.minLength(parseFloat(config.minLength))]);
        }
        // Add provided custom validators to the validator function
        if (config.validators) {
            config.validators.forEach(function (validatorConfig) {
                validator = forms.Validators.compose([validator, validatorConfig.validator]);
            });
        }
        return validator;
    };
    return TdDynamicFormsService;
}());
TdDynamicFormsService.decorators = [
    { type: core.Injectable },
];
/** @nocollapse */
TdDynamicFormsService.ctorParameters = function () { return []; };
/**
 * @param {?} parent
 * @return {?}
 */
function DYNAMIC_FORMS_PROVIDER_FACTORY(parent) {
    return parent || new TdDynamicFormsService();
}
var DYNAMIC_FORMS_PROVIDER = {
    // If there is already a service available, use that. Otherwise, provide a new one.
    provide: TdDynamicFormsService,
    deps: [[new core.Optional(), new core.SkipSelf(), TdDynamicFormsService]],
    useFactory: DYNAMIC_FORMS_PROVIDER_FACTORY,
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var noop$1 = function () {
    // empty method
};
var ELEMENT_INPUT_CONTROL_VALUE_ACCESSOR = {
    provide: forms.NG_VALUE_ACCESSOR,
    useExisting: core.forwardRef(function () { return TdDynamicElementComponent; }),
    multi: true,
};
var TdDynamicFormsErrorTemplate = /** @class */ (function (_super) {
    __extends(TdDynamicFormsErrorTemplate, _super);
    /**
     * @param {?} templateRef
     * @param {?} viewContainerRef
     */
    function TdDynamicFormsErrorTemplate(templateRef, viewContainerRef) {
        return _super.call(this, templateRef, viewContainerRef) || this;
    }
    return TdDynamicFormsErrorTemplate;
}(portal.TemplatePortalDirective));
TdDynamicFormsErrorTemplate.decorators = [
    { type: core.Directive, args: [{ selector: '[tdDynamicFormsError]ng-template' },] },
];
/** @nocollapse */
TdDynamicFormsErrorTemplate.ctorParameters = function () { return [
    { type: core.TemplateRef, },
    { type: core.ViewContainerRef, },
]; };
TdDynamicFormsErrorTemplate.propDecorators = {
    "tdDynamicFormsError": [{ type: core.Input },],
};
var TdDynamicElementDirective = /** @class */ (function () {
    /**
     * @param {?} viewContainer
     */
    function TdDynamicElementDirective(viewContainer) {
        this.viewContainer = viewContainer;
    }
    return TdDynamicElementDirective;
}());
TdDynamicElementDirective.decorators = [
    { type: core.Directive, args: [{
                selector: '[tdDynamicContainer]',
            },] },
];
/** @nocollapse */
TdDynamicElementDirective.ctorParameters = function () { return [
    { type: core.ViewContainerRef, },
]; };
var TdDynamicElementComponent = /** @class */ (function (_super) {
    __extends(TdDynamicElementComponent, _super);
    /**
     * @param {?} _componentFactoryResolver
     * @param {?} _dynamicFormsService
     */
    function TdDynamicElementComponent(_componentFactoryResolver, _dynamicFormsService) {
        var _this = _super.call(this) || this;
        _this._componentFactoryResolver = _componentFactoryResolver;
        _this._dynamicFormsService = _dynamicFormsService;
        /**
         * Sets label to be displayed.
         */
        _this.label = '';
        /**
         * Sets type or element of element to be rendered.
         * Throws error if does not exist or no supported.
         */
        _this.type = undefined;
        /**
         * Sets required validation checkup (if supported by element).
         */
        _this.required = undefined;
        /**
         * Sets min validation checkup (if supported by element).
         */
        _this.min = undefined;
        /**
         * Sets max validation checkup (if supported by element).
         */
        _this.max = undefined;
        /**
         * Sets minLength validation checkup (if supported by element).
         */
        _this.minLength = undefined;
        /**
         * Sets maxLength validation checkup (if supported by element).
         */
        _this.maxLength = undefined;
        /**
         * Sets selections for array elements (if supported by element).
         */
        _this.selections = undefined;
        _this.onModelChange = function (_) { return noop$1; };
        return _this;
    }
    Object.defineProperty(TdDynamicElementComponent.prototype, "value", {
        /**
         * @return {?}
         */
        get: function () {
            return this._value;
        },
        /**
         * @param {?} v
         * @return {?}
         */
        set: function (v) {
            if (v !== this._value) {
                this._value = v;
                this.onChange(v);
                this.onModelChange(v);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TdDynamicElementComponent.prototype, "maxAttr", {
        /**
         * @return {?}
         */
        get: function () {
            return this.max;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TdDynamicElementComponent.prototype, "minAttr", {
        /**
         * @return {?}
         */
        get: function () {
            return this.min;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TdDynamicElementComponent.prototype.ngOnInit = function () {
        var _this = this;
        var /** @type {?} */ ref = this._componentFactoryResolver.
            resolveComponentFactory(this._dynamicFormsService.getDynamicElement(this.type))
            .create(this.childElement.viewContainer.injector);
        this.childElement.viewContainer.insert(ref.hostView);
        this._instance = ref.instance;
        this._instance.control = this.dynamicControl;
        this._instance.label = this.label;
        this._instance.type = this.type;
        this._instance.value = this.value;
        this._instance.required = this.required;
        this._instance.min = this.min;
        this._instance.max = this.max;
        this._instance.minLength = this.minLength;
        this._instance.maxLength = this.maxLength;
        this._instance.selections = this.selections;
        this._instance.registerOnChange(function (value) {
            _this.value = value;
        });
        this.registerOnModelChange(function (value) {
            // fix to check if value is NaN (type=number)
            if (!Number.isNaN(value)) {
                _this._instance.value = value;
            }
        });
    };
    /**
     * Reassign any inputs that have changed
     * @param {?} changes
     * @return {?}
     */
    TdDynamicElementComponent.prototype.ngOnChanges = function (changes) {
        if (this._instance) {
            for (var /** @type {?} */ prop in changes) {
                this._instance[prop] = changes[prop].currentValue;
            }
        }
    };
    /**
     * Implemented as part of ControlValueAccessor.
     * @param {?} fn
     * @return {?}
     */
    TdDynamicElementComponent.prototype.registerOnModelChange = function (fn) {
        this.onModelChange = fn;
    };
    return TdDynamicElementComponent;
}(AbstractControlValueAccessor));
TdDynamicElementComponent.decorators = [
    { type: core.Component, args: [{
                providers: [TdDynamicFormsService, ELEMENT_INPUT_CONTROL_VALUE_ACCESSOR],
                selector: 'td-dynamic-element',
                template: '<div tdDynamicContainer></div>',
            },] },
];
/** @nocollapse */
TdDynamicElementComponent.ctorParameters = function () { return [
    { type: core.ComponentFactoryResolver, },
    { type: TdDynamicFormsService, },
]; };
TdDynamicElementComponent.propDecorators = {
    "dynamicControl": [{ type: core.Input },],
    "label": [{ type: core.Input },],
    "type": [{ type: core.Input },],
    "required": [{ type: core.Input },],
    "min": [{ type: core.Input },],
    "max": [{ type: core.Input },],
    "minLength": [{ type: core.Input },],
    "maxLength": [{ type: core.Input },],
    "selections": [{ type: core.Input },],
    "childElement": [{ type: core.ViewChild, args: [TdDynamicElementDirective,] },],
    "maxAttr": [{ type: core.HostBinding, args: ['attr.max',] },],
    "minAttr": [{ type: core.HostBinding, args: ['attr.min',] },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var TdDynamicFormsComponent = /** @class */ (function () {
    /**
     * @param {?} _formBuilder
     * @param {?} _dynamicFormsService
     * @param {?} _changeDetectorRef
     */
    function TdDynamicFormsComponent(_formBuilder, _dynamicFormsService, _changeDetectorRef) {
        this._formBuilder = _formBuilder;
        this._dynamicFormsService = _dynamicFormsService;
        this._changeDetectorRef = _changeDetectorRef;
        this._renderedElements = [];
        this._templateMap = new Map();
        this.dynamicForm = this._formBuilder.group({});
    }
    Object.defineProperty(TdDynamicFormsComponent.prototype, "elements", {
        /**
         * @return {?}
         */
        get: function () {
            return this._renderedElements;
        },
        /**
         * elements: ITdDynamicElementConfig[]
         * JS Object that will render the elements depending on its config.
         * [name] property is required.
         * @param {?} elements
         * @return {?}
         */
        set: function (elements) {
            if (elements) {
                this._elements = elements;
            }
            else {
                this._elements = [];
            }
            this._rerenderElements();
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TdDynamicFormsComponent.prototype, "form", {
        /**
         * Getter property for dynamic [FormGroup].
         * @return {?}
         */
        get: function () {
            return this.dynamicForm;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TdDynamicFormsComponent.prototype, "valid", {
        /**
         * Getter property for [valid] of dynamic [FormGroup].
         * @return {?}
         */
        get: function () {
            if (this.dynamicForm) {
                return this.dynamicForm.valid;
            }
            return false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TdDynamicFormsComponent.prototype, "value", {
        /**
         * Getter property for [value] of dynamic [FormGroup].
         * @return {?}
         */
        get: function () {
            if (this.dynamicForm) {
                return this.dynamicForm.value;
            }
            return {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TdDynamicFormsComponent.prototype, "errors", {
        /**
         * Getter property for [errors] of dynamic [FormGroup].
         * @return {?}
         */
        get: function () {
            if (this.dynamicForm) {
                var /** @type {?} */ errors = {};
                for (var /** @type {?} */ name in this.dynamicForm.controls) {
                    errors[name] = this.dynamicForm.controls[name].errors;
                }
                return errors;
            }
            return {};
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TdDynamicFormsComponent.prototype, "controls", {
        /**
         * Getter property for [controls] of dynamic [FormGroup].
         * @return {?}
         */
        get: function () {
            if (this.dynamicForm) {
                return this.dynamicForm.controls;
            }
            return {};
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TdDynamicFormsComponent.prototype.ngAfterContentInit = function () {
        this._updateErrorTemplates();
    };
    /**
     * Refreshes the form and rerenders all validator/element modifications.
     * @return {?}
     */
    TdDynamicFormsComponent.prototype.refresh = function () {
        this._rerenderElements();
        this._updateErrorTemplates();
    };
    /**
     * Getter method for error template references
     * @param {?} name
     * @return {?}
     */
    TdDynamicFormsComponent.prototype.getErrorTemplateRef = function (name) {
        return this._templateMap.get(name);
    };
    /**
     * Loads error templates and sets them in a map for faster access.
     * @return {?}
     */
    TdDynamicFormsComponent.prototype._updateErrorTemplates = function () {
        this._templateMap = new Map();
        for (var /** @type {?} */ i = 0; i < this._errorTemplates.toArray().length; i++) {
            this._templateMap.set(this._errorTemplates.toArray()[i].tdDynamicFormsError, this._errorTemplates.toArray()[i].templateRef);
        }
    };
    /**
     * @return {?}
     */
    TdDynamicFormsComponent.prototype._rerenderElements = function () {
        var _this = this;
        this._clearRemovedElements();
        this._renderedElements = [];
        var /** @type {?} */ duplicates = [];
        this._elements.forEach(function (elem) {
            _this._dynamicFormsService.validateDynamicElementName(elem.name);
            if (duplicates.indexOf(elem.name) > -1) {
                throw new Error("Dynamic element name: \"" + elem.name + "\" is duplicated");
            }
            duplicates.push(elem.name);
            if (!_this.dynamicForm.get(elem.name)) {
                _this.dynamicForm.addControl(elem.name, _this._dynamicFormsService.createFormControl(elem));
            }
            else {
                _this.dynamicForm.get(elem.name).setValidators(_this._dynamicFormsService.createValidators(elem));
            }
            // copy objects so they are only changes when calling this method
            _this._renderedElements.push(Object.assign({}, elem));
        });
        // call a change detection since the whole form might change
        this._changeDetectorRef.detectChanges();
        toPromise.toPromise.call(timer.timer()).then(function () {
            // call a markForCheck so elements are rendered correctly in OnPush
            _this._changeDetectorRef.markForCheck();
        });
    };
    /**
     * @return {?}
     */
    TdDynamicFormsComponent.prototype._clearRemovedElements = function () {
        var _this = this;
        for (var /** @type {?} */ i = 0; i < this._renderedElements.length; i++) {
            for (var /** @type {?} */ j = 0; j < this._elements.length; j++) {
                // check if the name of the element is still there removed
                if (this._renderedElements[i].name === this._elements[j].name) {
                    delete this._renderedElements[i];
                    break;
                }
            }
        }
        // remove elements that were removed from the array
        this._renderedElements.forEach(function (elem) {
            _this.dynamicForm.removeControl(elem.name);
        });
    };
    return TdDynamicFormsComponent;
}());
TdDynamicFormsComponent.decorators = [
    { type: core.Component, args: [{
                selector: 'td-dynamic-forms',
                template: "<form [formGroup]=\"dynamicForm\" novalidate>\n  <div class=\"td-dynamic-form-wrapper\">\n    <ng-template let-element ngFor [ngForOf]=\"elements\">\n      <div class=\"td-dynamic-element-wrapper\"\n          [style.max-width.%]=\"element.flex ? element.flex : 100\"\n          [style.flex]=\"'1 1 ' + (element.flex ? element.flex : 100) + '%'\"\n          [style.-ms-flex]=\"'1 1 ' + (element.flex ? element.flex : 100) + '%'\"\n          [style.-webkit-box-flex]=\"1\">\n        <td-dynamic-element\n          #dynamicElement\n          *ngIf=\"dynamicForm.controls[element.name]\"\n          [formControlName]=\"element.name\"\n          [dynamicControl]=\"dynamicForm.controls[element.name]\"\n          [id]=\"element.name\"\n          [label]=\"element.label || element.name\"\n          [type]=\"element.type\"\n          [required]=\"element.required\"\n          [min]=\"element.min\"\n          [max]=\"element.max\"\n          [minLength]=\"element.minLength\"\n          [maxLength]=\"element.maxLength\"\n          [selections]=\"element.selections\">\n        </td-dynamic-element>\n        <div class=\"tc-red-600\"\n             [style.font-size.%]=\"'70'\"\n             [style.position]=\"'absolute'\"\n             [style.bottom.px]=\"'10'\"\n              *ngIf=\"getErrorTemplateRef(element.name) && dynamicForm.controls[element.name]?.errors\">\n          <ng-template\n            [ngTemplateOutlet]=\"getErrorTemplateRef(element.name)\"\n            [ngTemplateOutletContext]=\"{control: dynamicForm.controls[element.name], errors: dynamicForm.controls[element.name]?.errors}\">\n          </ng-template>\n        </div>\n      </div>\n    </ng-template>\n  </div>\n  <ng-content></ng-content>\n</form>\n",
                styles: [".td-dynamic-form-wrapper{\n  -ms-flex-wrap:wrap;\n      flex-wrap:wrap;\n  -webkit-box-sizing:border-box;\n          box-sizing:border-box;\n  display:-webkit-box;\n  display:-ms-flexbox;\n  display:flex;\n  -webkit-box-orient:horizontal;\n  -webkit-box-direction:normal;\n      -ms-flex-direction:row;\n          flex-direction:row;\n  -webkit-box-align:center;\n      -ms-flex-align:center;\n          align-items:center;\n  -ms-flex-line-pack:center;\n      align-content:center;\n  max-width:100%;\n  -webkit-box-pack:start;\n      -ms-flex-pack:start;\n          justify-content:start; }\n  .td-dynamic-form-wrapper .td-dynamic-element-wrapper{\n    max-height:100%;\n    -webkit-box-sizing:border-box;\n            box-sizing:border-box;\n    position:relative;\n    padding:4px 4px 8px; }\n"],
                changeDetection: core.ChangeDetectionStrategy.OnPush,
            },] },
];
/** @nocollapse */
TdDynamicFormsComponent.ctorParameters = function () { return [
    { type: forms.FormBuilder, },
    { type: TdDynamicFormsService, },
    { type: core.ChangeDetectorRef, },
]; };
TdDynamicFormsComponent.propDecorators = {
    "_errorTemplates": [{ type: core.ContentChildren, args: [TdDynamicFormsErrorTemplate,] },],
    "elements": [{ type: core.Input, args: ['elements',] },],
};
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var TD_DYNAMIC_FORMS = [
    TdDynamicFormsComponent,
    TdDynamicElementComponent,
    TdDynamicElementDirective,
    TdDynamicFormsErrorTemplate,
];
var TD_DYNAMIC_FORMS_ENTRY_COMPONENTS = [
    TdDynamicInputComponent,
    TdDynamicFileInputComponent,
    TdDynamicTextareaComponent,
    TdDynamicSlideToggleComponent,
    TdDynamicCheckboxComponent,
    TdDynamicSliderComponent,
    TdDynamicSelectComponent,
    TdDynamicDatepickerComponent,
];
var CovalentDynamicFormsModule = /** @class */ (function () {
    function CovalentDynamicFormsModule() {
    }
    return CovalentDynamicFormsModule;
}());
CovalentDynamicFormsModule.decorators = [
    { type: core.NgModule, args: [{
                declarations: [
                    TD_DYNAMIC_FORMS,
                    TD_DYNAMIC_FORMS_ENTRY_COMPONENTS,
                ],
                imports: [
                    common.CommonModule,
                    forms.ReactiveFormsModule,
                    input.MatInputModule,
                    select.MatSelectModule,
                    checkbox.MatCheckboxModule,
                    slider.MatSliderModule,
                    slideToggle.MatSlideToggleModule,
                    icon.MatIconModule,
                    button.MatButtonModule,
                    datepicker.MatDatepickerModule,
                    common$1.CovalentCommonModule,
                    file.CovalentFileModule,
                ],
                exports: [
                    TD_DYNAMIC_FORMS,
                    TD_DYNAMIC_FORMS_ENTRY_COMPONENTS,
                ],
                providers: [
                    DYNAMIC_FORMS_PROVIDER,
                ],
                entryComponents: [TD_DYNAMIC_FORMS_ENTRY_COMPONENTS],
            },] },
];
/** @nocollapse */
CovalentDynamicFormsModule.ctorParameters = function () { return []; };

exports.CovalentDynamicFormsModule = CovalentDynamicFormsModule;
exports.TdDynamicFormsComponent = TdDynamicFormsComponent;
exports.ELEMENT_INPUT_CONTROL_VALUE_ACCESSOR = ELEMENT_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicFormsErrorTemplate = TdDynamicFormsErrorTemplate;
exports.TdDynamicElementDirective = TdDynamicElementDirective;
exports.TdDynamicElementComponent = TdDynamicElementComponent;
exports.TdDynamicType = TdDynamicType;
exports.TdDynamicElement = TdDynamicElement;
exports.DYNAMIC_ELEMENT_NAME_REGEX = DYNAMIC_ELEMENT_NAME_REGEX;
exports.TdDynamicFormsService = TdDynamicFormsService;
exports.DYNAMIC_FORMS_PROVIDER_FACTORY = DYNAMIC_FORMS_PROVIDER_FACTORY;
exports.DYNAMIC_FORMS_PROVIDER = DYNAMIC_FORMS_PROVIDER;
exports.TEXTAREA_INPUT_CONTROL_VALUE_ACCESSOR = TEXTAREA_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicTextareaComponent = TdDynamicTextareaComponent;
exports.SLIDER_INPUT_CONTROL_VALUE_ACCESSOR = SLIDER_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicSliderComponent = TdDynamicSliderComponent;
exports.SLIDE_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR = SLIDE_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicSlideToggleComponent = TdDynamicSlideToggleComponent;
exports.SELECT_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR = SELECT_TOGGLE_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicSelectComponent = TdDynamicSelectComponent;
exports.INPUT_INPUT_CONTROL_VALUE_ACCESSOR = INPUT_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicInputComponent = TdDynamicInputComponent;
exports.UPLOAD_INPUT_CONTROL_VALUE_ACCESSOR = UPLOAD_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicFileInputComponent = TdDynamicFileInputComponent;
exports.DATEPICKER_INPUT_CONTROL_VALUE_ACCESSOR = DATEPICKER_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicDatepickerComponent = TdDynamicDatepickerComponent;
exports.CHECKBOX_INPUT_CONTROL_VALUE_ACCESSOR = CHECKBOX_INPUT_CONTROL_VALUE_ACCESSOR;
exports.TdDynamicCheckboxComponent = TdDynamicCheckboxComponent;
exports.ɵa = AbstractControlValueAccessor;

Object.defineProperty(exports, '__esModule', { value: true });

})));
